Enabled war task

feenr opened the pull request:
Motivation:

In it's current state this project did not seem usable. There was no information in the readme or usage doc as to how it should be built or deployed.

Modification:
Added and configured the gradle war plugin to generate a deployable war file.

Reduced the number of Checkstyle warnings from ~6000 to ~60. These errors were primarily inclusion of tabs instead of spaces, non-conformant variable names,omitted brackets, and wildcard imports. The remaining warnings are primarily missing javadoc.

In the Travis CI configuration I have disabled checkstyle, findBugs and pmd. These will still run as part of the local build command.

Result:

I was successfully able to deploy to a tomcat server from the generated war file.
Checkstyle task still fails, but warnings are down significantly.

Fixes #4.

feenr committed:
Fix checkstyle warnings

feenr committed:
Enable war task

feenr committed:
Update travis.yaml to exclude failing tasks

feenr committed:
Correct unclosed UL in java doc

feenr committed:
Add war task to travis ci

feenr committed:
Moved war task to project specific build.gradle

cdancy commented:
Good job cleaning up the code!!! We're getting close to production level quality on at least one of our open source projects so there is that :)

cdancy commented:
And if you need with any of this don't be afraid to ask ;)

cdancy commented on .travis.yml:
```diff
@@ -13,7 +13,7 @@ cache:
 
 script:
-  - ./gradlew -s -Dorg.gradle.daemon=false clean build publishToMavenLocal
+  - ./gradlew -s -Dorg.gradle.daemon=false clean build -xcheckstyleMain -xfindBugsMain -xpmdMain war publishToMavenLocal
```
At the very least we should leave checkstyle if only to get a consistent look and feel throughout the codebase. The others I would just comment out in the `gradle/projects.gradle` file.

feenr replied:
Checkstyle has about 60 errors remaining. Almost all of them are missing javadoc. I'll disable these as you suggested.

cdancy commented on build.gradle:
```diff
@@ -1,32 +1,48 @@
         }
     }
+
+
```
Extra newline?

cdancy commented on build.gradle:
```diff
@@ -1,32 +1,48 @@
+dependencies {
+    warLibs 'org.ajoberstar:gradle-git:1.4.2'
+    warLibs 'net.ltgt.gradle:gradle-errorprone-plugin:0.0.13'
+    warLibs 'com.github.jengelman.gradle.plugins:shadow:2.0.1'
```
You're adding these to the war libs? These are meant for buildscript execution only and should not be part of anything we ship.

feenr replied:
You are correct. These should not be added.

cdancy commented on build.gradle:
```diff
@@ -37,12 +53,19 @@ release {
+war {
+    webInf { from "$rootDir//main/java" }
+    classpath configurations.warLibs
+}
```
I would add this code to the `gradle/projects.gradle` file. The root file is meant to serve as the "release process" so to speak.

feenr replied:
Ah okay. I had it there originally and then moved it in my latest commit. My thinking was that projects.gradle appeared to be gradle tasks common among all subprojects (even though this repo only has one project in it). If a new subproject were added, I wouldn't want it to automatically inherit the war task as it would require project specific configuration.

cdancy replied:
That file has an `allprojects` closure that will take effect for everything. 

Yeah I don't think we'll be adding a sub-project here so it's a safe place to put things ;)

feenr replied:
Does that mean you agree with my reasoning? Or would you still recommend moving to `projects.gradle`?

cdancy replied:
hahaha ... no move to `projects.gradle`

feenr commented:
@cdancy One thing I encountered while cleaning up the project is that the errorprone plugin was getting an unhandled exception parsing one of the files.

I ended up disabling it from the build arguments.

cdancy commented:
@feenr that's sad as it's a good tool to have to keep code clean and avoid potential issues. Do you have the issue/exception still around?

feenr committed:
Applied code review comments.

feenr commented:
@cdancy I addressed most of your concerns.
After fixing checkstyle issues, the parsing exception from errorprone seems to have gone away. I still have the -WError flag removed so that it can build successfully. 

I'll try to fix all of those compile time warnings before merging this branch. I also also need to get checkstyle task passing if we're going to leave that enabled.

What needs to be done to get the war published to bintray? Is that something you can help with?

cdancy commented:
@feenr once things are ready I can certainly help with getting things published to bintray if that's what's required.

feenr committed:
Fix all checkstyle warnings

feenr committed:
Fix compile warnings. Enable error on compile warn

cdancy commented on gradle/projects.gradle:
```diff
@@ -32,16 +36,33 @@ allprojects {
+        testCompile ('org.assertj:assertj-core:3.8.0')
+        warLibs 'org.ajoberstar:gradle-git:1.4.2'
+        warLibs 'net.ltgt.gradle:gradle-errorprone-plugin:0.0.13'
+        warLibs 'com.github.jengelman.gradle.plugins:shadow:2.0.1'
```
Are the above 3 still needed?
